tests/e2e/metrics/test_metrics.py

```diff
@@ -1,40 +1,69 @@
-import datetime
-import uuid
+import json
 
-import boto3
 import pytest
-from e2e import conftest
-from e2e.utils import helpers
 
+from tests.e2e.utils import helpers
 
-@pytest.fixture(scope="module")
-def config() -> conftest.LambdaConfig:
-    return {
-        "parameters": {},
-        "environment_variables": {
-            "POWERTOOLS_METRICS_NAMESPACE": "powertools-e2e-metric",
-            "POWERTOOLS_SERVICE_NAME": "test-powertools-service",
-            "METRIC_NAME": f"business-metric-{str(uuid.uuid4()).replace('-','_')}",
-        },
-    }
 
+@pytest.fixture
+def basic_handler_fn(infrastructure: dict) -> str:
+    return infrastructure.get("BasicHandler", "")
 
-def test_basic_lambda_metric_visible(execute_lambda: conftest.InfrastructureOutput, config: conftest.LambdaConfig):
+
+@pytest.fixture
+def basic_handler_fn_arn(infrastructure: dict) -> str:
+    return infrastructure.get("BasicHandlerArn", "")
+
+
+@pytest.fixture
+def cold_start_fn(infrastructure: dict) -> str:
+    return infrastructure.get("ColdStart", "")
+
+
+@pytest.fixture
+def cold_start_fn_arn(infrastructure: dict) -> str:
+    return infrastructure.get("ColdStartArn", "")
+
+
+METRIC_NAMESPACE = "powertools-e2e-metric"
+
+
+def test_basic_lambda_metric_is_visible(basic_handler_fn: str, basic_handler_fn_arn: str):
     # GIVEN
-    start_date = execute_lambda.get_lambda_execution_time()
-    end_date = start_date + datetime.timedelta(minutes=5)
+    metric_name = helpers.build_metric_name()
+    service = helpers.build_service_name()
+    dimensions = helpers.build_add_dimensions_input(service=service)
+    metrics = helpers.build_multiple_add_metric_input(metric_name=metric_name, value=1, quantity=3)
 
     # WHEN
+    event = json.dumps({"metrics": metrics, "service": service, "namespace": METRIC_NAMESPACE})
+    _, execution_time = helpers.trigger_lambda(lambda_arn=basic_handler_fn_arn, payload=event)
+
+    metrics = helpers.get_metrics(
```

Comment:
I like the idea of having fully independent tests that calls lambda with dynamic input. Not using cache simplifies things, but also I see a risk here. If we imagine we have 20-30 or even more tests in total this will lead to having the same number of parallelized processes calling lambdas and waiting in worst case scenario 1,2 minutes on a data being present in AWS Service ( and retrying for a result every 5 seconds). It may cause resource saturation issues, especially on limited CI environment + we may hit API throttling issue for CW in get_metric or other methods that fetch the result . Using cache, gave us the same performance but without hard parallelization needed. Do we have any data validating/invalidating my theory?

Reply:
I totally see your concern and all great points - allow me to expand on why our risk is minimized as we spoke earlier. 

Caching wouldn't help as the payload is now dynamic to ease writing handlers and more explicit tests, it could cause the opposite problem now where we're filling out memory (not really a problem, just a data point) and no cache hit.

For the resource exhaustion, we have a **controllable demand** where we own when those tests will run (on merge/push and on label add). This reduces the risk. We also have jitter and delays that can adapt retries based on their APIs they might be exhausting - this means we would even fail the test before we are able to succeed OR we succeed eventually within our max retry. 

For customers' laptop resource exhaustion, it is also minimized because **(1)** we control the recommendation on how they should run (one vs one module vs the entire test suite), **(2)** E2E will cost customers money thus giving them an incentive to run at our CI instead, and **(3)** E2E tests will run on a single runtime compared to CI (3.7, 3.8, 3.9 etc.).